### PR TITLE
Task 5 implementation

### PR DESCRIPTION
**Pull request by maksumov**
# Task 5

### Prerequisites

- [x] The task is a continuation of Homework 4 and should be done in the same repos
- [x] Install the latest version of AWS SDK (https://aws.amazon.com/ru/sdk-for-node-js/)
- [x] Install the CSV parser package (https://www.npmjs.com/package/csv-parser)
- [x] Sign in into AWS Console and create a new S3 bucket with a folder, called uploaded
- [x] Pull latest changes from nodejs-aws-fe (https://github.com/rolling-scopes-school/nodejs-aws-fe)

### Task 5.1

- [x] Create a new service, called import-service, with a its own serverless.yml file at the same level as product service.
- [x] The backend project structure should look like:
  - src/product-service
  - src/import-service
- [x] Create a lambda function in that serverless.yml file, called importProductsFile, which will be triggered by the HTTP GET method.
- [x] The requested URL should be /import.
- [x] The name of CSV file with products will be passed in a query string as a name parameter and should be described in the serverless.yml file as a request parameter.
- [x] Update serverless.yml with policies to allow lambda functions to interact with S3.
- [x] A new Signed URL should be created with the following Key: uploaded/${fileName}.
- [x] The response from the lambda should be the created Signed URL.
- [x] The lambda endpoint should be integrated with the frontend by updating import property of the API paths configuration.

### Task 5.2

- [x] Create a lambda function under the same serverless.yml file, called importFileParser, which will be triggered by the S3 s3:ObjectCreated:* event.
- [x] Configure the event to be triggered only by changes in the uploaded folder in S3.
- [x] The lambda function should use the Readable stream to get an object from S3, parse it via csv-parser and log each record to be shown in CloudWatch.
- [x] The response should be a correct HTTP response code.

Source: https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task5-import-to-s3/task.md

## Task 5 Solution URLs (planned)

- App FE:
  - Main page: https://dl0k87cgjwcxa.cloudfront.net/
  - Admin products page: https://dl0k87cgjwcxa.cloudfront.net/admin/products
  - Pull Request URL: https://github.com/maksumov/nodejs-aws-fe/pull/5
- API Endpoints:
  - Product Service: https://wnxy68e0x6.execute-api.eu-west-1.amazonaws.com/dev/products
  - Import Service: https://6s6jrzco26.execute-api.eu-west-1.amazonaws.com/dev/import
- ...

### Main Tasks

- [x] File serverless.yml contains configuration for importProductsFile function
- [x] The importProductsFile lambda function returns a correct response which can be used to upload a file into the S3 bucket
- [x] Frontend application is integrated with importProductsFile lambda
- [x] The importFileParser lambda function is implemented and serverless.yml contains configuration for the lambda

### Additional Tasks

- [x] async/await is used in lambda functions
- [ ] importProductsFile lambda is covered by unit tests (aws-sdk-mock can be used to mock S3 methods - https://www.npmjs.com/package/aws-sdk-mock)
- [x] At the end of the stream the lambda function should move the file from the uploaded folder into the parsed folder (move the file means that file should be copied into parsed folder, and then deleted from uploaded folder)

### CODE REVIEW

File: src/import-service/handlers/importProductsFile.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import logger from "./utils/logger";
+import { response } from "./utils/response";
+
+const importProductsFile: APIGatewayProxyHandler = async (
+  event: APIGatewayProxyEvent
+) => {
+  logger.info(event, "importProductsFile");
+
+  // No filename or bad filename
+  if (
```
Comment (mmarkelov) [inline src/import-service/handlers/importProductsFile.ts]: You can use optional chaining to reduce your code
```js
if (!/^[\w,\s-]+\.[A-Za-z]{3}$/.test(event.queryStringParameters?.name))
```

File: src/import-service/serverless.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    iamRoleStatements: [
+      {
+        Effect: "Allow",
+        Action: "s3:ListBucket",
+        Resource: ["arn:aws:s3:::maksumov-bucket-for-images"],
+      },
+      {
+        Effect: "Allow",
+        Action: ["s3:*"],
+        Resource: ["arn:aws:s3:::maksumov-bucket-for-images/*"],
```
Comment (mmarkelov) [inline src/import-service/serverless.ts]: It's better to put this in some variable to change it easily

**Comment by mmarkelov** [comment]
@maksumov are you planning to add tests for **importProductsFile**?

File: src/import-service/serverless.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    name: "import-service",
+    // app and org for use with dashboard.serverless.com
+    // app: your-app-name,
+    // org: your-org-name,
+  },
+  frameworkVersion: "2",
+  custom: {
+    webpack: {
+      webpackConfig: "./webpack.config.js",
+      includeModules: true,
```
Comment (mmarkelov) [inline src/import-service/serverless.ts:16]: By the way you can reduce your bundle size with it
```js
includeModules: {
    forceExclude: ['aws-sdk']
}
```

**Comment by maksumov** [comment]
> @maksumov are you planning to add tests for **importProductsFile**?

I plan to add tests. Yesterday I was stuck on a simple task, today I plan to finish all the tasks to the end if possible. You can give me a score based on the current situation and if I add new functionality and if you have time - increase my score.

**Pull request merged**